bin/runtime-common/src/messages_batch_extension.rs

```diff
@@ -0,0 +1,297 @@
+		_who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		Ok(ValidTransaction::default())
+	}
+
+	fn pre_dispatch(
+		self,
+		_who: &Self::AccountId,
+		_call: &Self::Call,
+		_post_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		// TODO: for every call from the batch - call the `BridgesExtension` to ensure that every
+		// call transaction brings something new and reject obsolete transactions
+		unimplemented!("TODO") // TODO: return actual call type
+	}
+
+	fn post_dispatch(
+		pre: Option<Self::Pre>,
+		info: &DispatchInfoOf<Self::Call>,
+		post_info: &PostDispatchInfoOf<Self::Call>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		// we never refund anything if that is not bridge transaction or if it is a bridge
+		// transaction that we do not support here
+		let (relayer, call_type) = match pre {
+			Some(pre) => (pre.relayer, pre.call_type),
+			None => return Ok(()),
+		};
+
+		// we never refund anything if transaction has failed
+		if result.is_err() {
+			return Ok(())
+		}
+
+		// check if relay chain state has been updated
+		if let CallType::AllFinalityAndDelivery(expected_relay_chain_state, _, _) = call_type {
+			let actual_relay_chain_state = relay_chain_state::<Runtime, GrandpaInstance>();
+			if actual_relay_chain_state != expected_relay_chain_state {
+				// we only refund relayer if all calls have updated chain state
+				return Ok(())
+			}
+
+			// there's a conflict between how bridge GRANDPA pallet works and the
+			// `AllFinalityAndDelivery` transaction. If relay cahin header is mandatory, the GRANDPA
+			// pallet returns `Pays::No`, because such transaction is mandatory for operating the
+			// bridge. But `utility.batchAll` transaction always requires payment. But in both cases
+			// we'll refund relayer - either explicitly here, or using `Pays::No` if he's choosing
+			// to submit dedicated transaction.
+		}
+
+		// check if parachain state has been updated
+		match call_type {
+			CallType::AllFinalityAndDelivery(_, expected_parachain_state, _) |
+			CallType::ParachainFinalityAndDelivery(expected_parachain_state, _) => {
+				let actual_parachain_state = parachain_state::<Runtime, ParachainsInstance>();
+				if expected_parachain_state != actual_parachain_state {
+					// we only refund relayer if all calls have updated chain state
+					return Ok(())
+				}
+			},
+			_ => (),
+		}
+
+		// check if messages have been delivered
+		let actual_messages_state = messages_state::<Runtime, MessagesInstance>();
+		let pre_dispatch_messages_state = call_type.pre_dispatch_messages_state();
+		if actual_messages_state == pre_dispatch_messages_state {
+			// we only refund relayer if all calls have updated chain state
+			return Ok(())
+		}
+
+		// regarding the tip - refund that happens here (at this side of the bridge) isn't the whole
```

Comment:
That's an important thing - we don't refund the whole transaction cost here. Tip is not included in refund. Reasoning is in the code